modules/gapi/src/compiler/gasync.cpp

```diff
@@ -0,0 +1,136 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
+
+#include "opencv2/gapi/gcomputation_async.hpp"
+#include "opencv2/gapi/gcomputation.hpp"
+#include "opencv2/gapi/gcompiled_async.hpp"
+#include "opencv2/gapi/gcompiled.hpp"
+
+#include <condition_variable>
+
+#include <future>
+#include <condition_variable>
+//#include <chrono>
+#include <stdexcept>
+#include <queue>
+
+namespace {
+    //This is a tool to to move initialize captures of a lambda in C++11
+    template<typename T>
+    struct move_through_copy{
+       T value;
+       move_through_copy(T&& g) : value(std::move(g)) {}
+       move_through_copy(move_through_copy&&) = default;
+       move_through_copy(move_through_copy const& lhs) : move_through_copy(std::move(const_cast<move_through_copy&>(lhs))) {}
+    };
+}
+
+namespace cv {
+
+    namespace impl{
+        class async_service {
+
+            std::mutex mtx;
+            std::condition_variable cv;
+            std::queue<std::function<void()>> q;
+            std::atomic<bool> exiting;
+
+            std::thread thrd;
+
+        public:
+            async_service() = default ;
+
+            void add_task(std::function<void()>&& t){
+                if (!thrd.joinable())
+                {
+                    //thread has not been started yet, so start it
+                    thrd = std::thread {[this](){
+                        auto second_q {q};
+                        while (!exiting){
+                            std::unique_lock<std::mutex> lck{mtx};
+                            if (q.empty())
+                            {
+                                cv.wait(lck, [&](){ return exiting || !q.empty();});
+                            }
+
+                            q.swap(second_q);
+                            lck.unlock();
+
+                            if (!second_q.empty())
+                            {
+                                auto f = second_q.front();
+                                second_q.pop();
+                                f();
+                            }
+                        }
+                    }};
+
+                }
+                std::unique_lock<std::mutex> lck{mtx};
+                bool first_task = q.empty();
+                q.push(std::move(t));
+                lck.unlock();
+                if (first_task)
+                {
+                    cv.notify_one();
+                }
+            }
+            ~async_service(){
+                if (thrd.joinable())
+                {
+                    exiting = true;
+                    mtx.lock();
+                    mtx.unlock();
+                    cv.notify_one();
+                    thrd.join();
```

Comment:
Be careful with waiting for notifications from worker threads during "cleanup".
Win32 `ExitProcess()` terminates all non-main threads **unconditionally** - probably with broken synchronization state - holding mutexes (user space at least) and other synchronization primitives.
And after that, it calls `atexit()` handlers from shared libraries, like this destructor.

Hopefully `thrd.joinable()` can handle that properly.